lib/utils/device/live.ts

```diff
@@ -0,0 +1,409 @@
+				const dockerfile = buildTask.dockerfile;
+				if (dockerfile == null) {
+					throw new Error(
+						`Could not detect dockerfile for service: ${serviceName}`,
+					);
+				}
+
+				if (buildTask.dockerfilePath == null) {
+					// this is a bit of a hack as resin-bundle-resolve
+					// does not always export the dockerfilePath, this
+					// only happens when the dockerfile path is
+					// specified differently - this should be patched
+					// in resin-bundle-resolve
+					this.dockerfilePaths[
+						buildTask.serviceName
+					] = this.getDockerfilePathFromTask(buildTask);
+				} else {
+					this.dockerfilePaths[buildTask.serviceName] = [
+						buildTask.dockerfilePath,
+					];
+				}
+
+				// Find the containerId from the device state
+				const container = _.find(this.lastDeviceStatus!.containers, {
+					serviceName,
+				});
+				if (container == null) {
+					throw new Error(
+						`Could not find a container on device for service: ${serviceName}`,
+					);
+				}
+
+				const log = (msg: string) => {
+					this.logger.logLivepush(`[service ${serviceName}] ${msg}`);
+				};
+
+				const livepush = await Livepush.init(
+					dockerfile,
+					context,
+					container.containerId,
+					this.imageIds[serviceName],
+					this.docker,
+				);
+
+				livepush.on('commandExecute', command =>
+					log(`Executing command: \`${command.command}\``),
+				);
+				livepush.on('commandOutput', output =>
+					log(`   ${output.output.data.toString()}`),
+				);
+
+				// TODO: Memoize this for containers which share a context
+				const monitor = chokidar.watch('.', {
+					cwd: context,
+					ignoreInitial: true,
+				});
+				monitor.on('add', (changedPath: string) =>
+					this.handleFSEvent({
+						filename: changedPath,
+						type: 'add',
+						serviceName,
+					}),
+				);
+				monitor.on('change', (changedPath: string) =>
+					this.handleFSEvent({
+						filename: changedPath,
+						type: 'change',
+						serviceName,
+					}),
+				);
+				monitor.on('unlink', (changedPath: string) =>
+					this.handleFSEvent({
+						filename: changedPath,
+						type: 'unlink',
+						serviceName,
+					}),
+				);
+				this.containers[serviceName] = {
+					livepush,
+					context,
+					monitor,
+					containerId: container.containerId,
+				};
+			}
+		}
+
+		// Setup cleanup handlers for the device
+
+		// This is necessary because the `exit-hook` module is used by several
+		// dependencies, and will exit without calling the following handler.
+		// Once https://github.com/balena-io/balena-cli/issues/867 has been solved,
+		// we are free to (and definitely should) remove the below line
+		process.removeAllListeners('SIGINT');
+		process.on('SIGINT', async () => {
+			this.logger.logLivepush('Cleaning up device...');
+			await Promise.all(
+				_.map(this.containers, container => {
+					container.livepush.cleanupIntermediateContainers();
+				}),
+			);
+
+			process.exit(0);
```

Comment:
`process.exitCode = 0` instead maybe?

Reply:
We need to actually force an exit, as we caught the interrupt.

Reply:
Ah, good point.